### PR TITLE
Enhancement: Add project id parameter

### PR DESCRIPTION
**Pull request by patricklubach**
# Pull Request Template

## Description

I needed to add a the project_id variable as I recognized that it is only possible to pass in the project id when the Google provider is configured through a variable. But I ran into a scenario where the project id comes from a terraform module that is not passable to the Google provider because this cause a circle of dependencies. So I introduced the project id variable to make it possible. The support for retrieving the project id through the Google provider is still there, but now the project_id variable takes precedence over the current implementation. The current implementation acts now as a fallback if the project id is not passed in as a variable.

Fixes # (issue)

## Type of change

- [x] New feature (non-breaking change which adds functionality) / Enhancement
- [x] This change requires a documentation update

## How Has This Been Tested?

This was tested with the minimal example in `examples/minimal` directory.

- [x] Passed in no project id (nor through provider, neither by variable). Expectation: failure | Result: failure
- [x] Passed in project id by provider only. Expectation: success | Result: success
- [x] Passed in project id by variable only. Expectation: success | Result: success
- [x] Passed in project id by provider and by variable. Expectation: success (variable value was taken) | Result: success

## Checklist:

- [x] My PR title follows [semantics prefix](.github/CONTRIBUTING.md#semantic-pull-requests)
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked my code and corrected any misspellings
- [x] I have run pre-commit hooks `pre-commit run -a`
- [x] CI tests are passing

### CODE REVIEW

**Comment by patricklubach** [comment]
@anouarchattouna @PascalBourdier would you be so kind and review? :)

**Comment by patricklubach** [comment]
Please someone of @anouarchattouna @PascalBourdier @mirakl-admin can you review?

**Comment by patricklubach** [comment]
Does anyone has time for review? If not, I will continue developing it on my own.

File: locals.tf [hunk 1]
```diff
@@ -1,6 +1,6 @@
 locals {
   # get project id (used by google provider)
```
Comment (anouarchattouna) [inline locals.tf:2]:
```suggestion
  # get project id: either from project_id variable or from google provider.
```

File: variables.tf [hunk 2]
```diff
@@ -1,3 +1,9 @@
+variable "project_id" {
+  description = "GCP project that the Cloud Run instances are being created."
```
Comment (anouarchattouna) [inline variables.tf:2]:
```suggestion
  description = "The ID of the project to which the resource belongs. If it is not provided, the provider project is used."
```

**Review by anouarchattouna — CHANGES_REQUESTED** [review]

**Review by madhifallah — APPROVED** [review]
LGTM

**Comment by patricklubach** [comment]
@anouarchattouna I implement you requested changes. I don't get what is wrong now with the CI. Can you explain to me please?

**Review by anouarchattouna — APPROVED** [review]

**Comment by anouarchattouna** [comment]
> @anouarchattouna I implement you requested changes. I don't get what is wrong now with the CI. Can you explain to me please?

Please run pre-commit again

**Comment by patricklubach** [comment]
When running pre-commit I get the following output:

`
$ pre-commit run -a
Terraform fmt............................................................Passed
Terraform validate.......................................................Failed
- hook id: terraform_validate
- exit code: 1

Validation failed: examples/complete
╷
│ Warning: Experimental feature "module_variable_optional_attrs" is active
│ 
│   on ../../versions.tf line 3, in terraform:
│    3:   experiments      = [module_variable_optional_attrs]
│ 
│ Experimental features are subject to breaking changes in future minor or
│ patch releases, based on feedback.
│ 
│ If you have feedback on the design of this feature, please open a GitHub
│ issue to discuss it.
╵
╷
│ Error: Could not load plugin
│ 
│ 
│ Plugin reinitialization required. Please run "terraform init".
│ 
│ Plugins are external binaries that Terraform uses to access and manipulate
│ resources. The configuration provided requires plugins which can't be
│ located,
│ don't satisfy the version constraints, or are otherwise incompatible.
│ 
│ Terraform automatically discovers provider requirements from your
│ configuration, including providers used in child modules. To see the
│ requirements and constraints, run "terraform providers".
│ 
│ 2 problems:
│ 
│ - failed to instantiate provider "registry.terraform.io/hashicorp/google"
│ to obtain schema: there is no package for
│ registry.terraform.io/hashicorp/google 4.9.0 cached in .terraform/providers
│ - failed to instantiate provider
│ "registry.terraform.io/hashicorp/google-beta" to obtain schema: there is no
│ package for registry.terraform.io/hashicorp/google-beta 4.9.0 cached in
│ .terraform/providers
│ 
╵

Terraform docs...........................................................Failed
- hook id: terraform_docs
- files were modified by this hook
Terraform validate with tfsec............................................Passed
Terraform validate with tflint...........................................Passed
check for merge conflicts................................................Passed
`

When I run `terraform init` and `terraform validate` everything is fine and it succeeds. Does anyone know what's the cause?

**Comment by patricklubach** [comment]
> When running pre-commit I get the following output:
> 
> ```
> $ pre-commit run -a
> Terraform fmt............................................................Passed
> Terraform validate.......................................................Failed
> - hook id: terraform_validate
> - exit code: 1
> 
> Validation failed: examples/complete
> ╷
> │ Warning: Experimental feature "module_variable_optional_attrs" is active
> │ 
> │   on ../../versions.tf line 3, in terraform:
> │    3:   experiments      = [module_variable_optional_attrs]
> │ 
> │ Experimental features are subject to breaking changes in future minor or
> │ patch releases, based on feedback.
> │ 
> │ If you have feedback on the design of this feature, please open a GitHub
> │ issue to discuss it.
> ╵
> ╷
> │ Error: Could not load plugin
> │ 
> │ 
> │ Plugin reinitialization required. Please run "terraform init".
> │ 
> │ Plugins are external binaries that Terraform uses to access and manipulate
> │ resources. The configuration provided requires plugins which can't be
> │ located,
> │ don't satisfy the version constraints, or are otherwise incompatible.
> │ 
> │ Terraform automatically discovers provider requirements from your
> │ configuration, including providers used in child modules. To see the
> │ requirements and constraints, run "terraform providers".
> │ 
> │ 2 problems:
> │ 
> │ - failed to instantiate provider "registry.terraform.io/hashicorp/google"
> │ to obtain schema: there is no package for
> │ registry.terraform.io/hashicorp/google 4.9.0 cached in .terraform/providers
> │ - failed to instantiate provider
> │ "registry.terraform.io/hashicorp/google-beta" to obtain schema: there is no
> │ package for registry.terraform.io/hashicorp/google-beta 4.9.0 cached in
> │ .terraform/providers
> │ 
> ╵
> 
> Terraform docs...........................................................Failed
> - hook id: terraform_docs
> - files were modified by this hook
> Terraform validate with tfsec............................................Passed
> Terraform validate with tflint...........................................Passed
> check for merge conflicts................................................Passed
> ```
> 
> When I run `terraform init` and `terraform validate` everything is fine and it succeeds. Does anyone know what's the cause?

okay there was an issue with the .terraform folder in `examples/complete`. Removed it, ran pre-commit again and it passed.

**Pull request closed without merging**